storage/local/doubledelta.go

```diff
@@ -0,0 +1,513 @@
+// timestamps calculated by applying the base value and time and the base deltas.
+// However, once 8 bytes would be needed to encode a double-delta value, a
+// fall-back to the absolute numbers happens (so that timestamps are saved
+// directly as int64 and values as float64).
+// doubleDeltaEncodedChunk implements the chunk interface.
+type doubleDeltaEncodedChunk []byte
+
+// newDoubleDeltaEncodedChunk returns a newly allocated doubleDeltaEncodedChunk.
+func newDoubleDeltaEncodedChunk(tb, vb deltaBytes, isInt bool, length int) *doubleDeltaEncodedChunk {
+	if tb < 1 {
+		panic("need at least 1 time delta byte")
+	}
+	if length < doubleDeltaHeaderBytes+16 {
+		panic(fmt.Errorf(
+			"chunk length %d bytes is insufficient, need at least %d",
+			length, doubleDeltaHeaderBytes+16,
+		))
+	}
+	c := make(doubleDeltaEncodedChunk, doubleDeltaHeaderIsIntOffset+1, length)
+
+	c[doubleDeltaHeaderTimeBytesOffset] = byte(tb)
+	c[doubleDeltaHeaderValueBytesOffset] = byte(vb)
+	if vb < d8 && isInt { // Only use int for fewer than 8 value double-delta bytes.
+		c[doubleDeltaHeaderIsIntOffset] = 1
+	} else {
+		c[doubleDeltaHeaderIsIntOffset] = 0
+	}
+	return &c
+}
+
+func (c doubleDeltaEncodedChunk) newFollowupChunk() chunk {
+	return newDoubleDeltaEncodedChunk(d1, d0, true, cap(c))
+}
+
+func (c doubleDeltaEncodedChunk) baseTime() clientmodel.Timestamp {
+	return clientmodel.Timestamp(
+		binary.LittleEndian.Uint64(
+			c[doubleDeltaHeaderBaseTimeOffset:],
+		),
+	)
+}
+
+func (c doubleDeltaEncodedChunk) baseValue() clientmodel.SampleValue {
+	return clientmodel.SampleValue(
+		math.Float64frombits(
+			binary.LittleEndian.Uint64(
+				c[doubleDeltaHeaderBaseValueOffset:],
+			),
+		),
+	)
+}
+
+func (c doubleDeltaEncodedChunk) baseTimeDelta() clientmodel.Timestamp {
+	return clientmodel.Timestamp(
+		binary.LittleEndian.Uint64(
+			c[doubleDeltaHeaderBaseTimeDeltaOffset:],
+		),
+	)
+}
+
+func (c doubleDeltaEncodedChunk) baseValueDelta() clientmodel.SampleValue {
+	return clientmodel.SampleValue(
+		math.Float64frombits(
+			binary.LittleEndian.Uint64(
+				c[doubleDeltaHeaderBaseValueDeltaOffset:],
+			),
+		),
+	)
+}
+
+func (c doubleDeltaEncodedChunk) timeBytes() deltaBytes {
+	return deltaBytes(c[doubleDeltaHeaderTimeBytesOffset])
+}
+
+func (c doubleDeltaEncodedChunk) valueBytes() deltaBytes {
+	return deltaBytes(c[doubleDeltaHeaderValueBytesOffset])
+}
+
+func (c doubleDeltaEncodedChunk) sampleSize() int {
+	return int(c.timeBytes() + c.valueBytes())
+}
+
+func (c doubleDeltaEncodedChunk) len() int {
+	if len(c) <= doubleDeltaHeaderIsIntOffset+1 {
+		return 0
+	}
+	if len(c) <= doubleDeltaHeaderBaseValueOffset+8 {
+		return 1
+	}
+	return (len(c)-doubleDeltaHeaderBytes)/c.sampleSize() + 2
+}
+
+func (c doubleDeltaEncodedChunk) isInt() bool {
+	return c[doubleDeltaHeaderIsIntOffset] == 1
+}
+
+// add implements chunk.
+func (c doubleDeltaEncodedChunk) add(s *metric.SamplePair) []chunk {
+	if len(c) <= doubleDeltaHeaderIsIntOffset+1 {
```

Comment:
Just noticed that this could be `if c.len() == 0`. Same for `deltaEncodedChunk.add()`.

Shorter and clearer.